lib/services/role.go

```diff
@@ -1511,6 +1520,52 @@ func (set RoleSet) CheckAccessToDatabase(server types.DatabaseServer, dbName, db
 	return trace.AccessDenied("access to database denied")
 }
 
+// CheckAccessToDatabaseUser checks if this role set is allowed to connect to
+// a database as a particular database account.
+//
+// Used as an authorization check when connecting to a database.
+func (set RoleSet) CheckAccessToDatabaseUser(server types.DatabaseServer, dbUser string) error {
+	var errs []error
+	// Check deny rules.
+	for _, role := range set {
+		matchNamespace, namespaceMessage := MatchNamespace(role.GetNamespaces(Deny), server.GetNamespace())
+		matchLabels, labelsMessage, err := MatchLabels(role.GetDatabaseLabels(Deny), server.GetAllLabels())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		matchUser, userMessage := MatchDatabaseUser(role.GetDatabaseUsers(Deny), dbUser)
+		if matchNamespace && matchLabels && matchUser {
```

Comment:
Is this intention? For example for servers its:

```go
if matchNamespace && (matchLabels || matchLogin) {
```

Reply:
Good question, it actually is intentional, because I was thinking for example about a scenario where you may want to deny access to a particular database account, say, in a production database, so you would do something like:

```yaml
deny:
  db_labels:
    environment: ["prod"]
  db_users: ["postgres"]
```

With the OR condition here, it would get greedy and prohibit access to any database with "prod" label, and to "postgres" account in any database, which is a bit unexpected behavior, at least to me.

That said, you're right that SSH access check does that "greedy" matching - which I think makes implementing the scenario similar to what I mentioned above quite hard (i.e. "I want to deny SSH 'root' login to specific set of servers"). I also find it quite weird that the labels/logins are `&&`'ed in the "allow" section and `||`'ed in the "deny" section.

So yeah, this is how I intended it to work, however I'm open to changing this if you feel like this is wrong or I'm missing something.

Reply:
I think the intention with Teleport is that you should start with zero access, then gradually grant access by `allow` directives. After that, `deny` rules will _always_ take precedence over allow and will match greedily. Essentially RBAC rules should generally have many `allow` directives and 0-1 `deny` directives - using `deny` should be a last resort for a situation where you want a certain role to NEVER have access to a particular user OR label under any circumstances.

I think that the way to implement the proposed solution from your comment would be to give the people who shouldn't have production access a role which specifically **doesn't** have this directive i.e:

```yaml
allow:
  db_users: ['postgres']
  db_labels:
    environment: ("dev|staging")
```

Alternatively, you add a label to every non-prod database and then grant access based on that. I do find this a little confusing and counter-intuitive myself, having recently started creating roles and giving access to people via the infra clusters - but for now we should copy the same behaviour.

Reply:
@r0mant Making Database Access RBAC behave differently does increase the complexity the user has to keep in their head. For example, I can see cut/paste errors happen quite easily here if someone wants to copy similar RBAC rules from Server Access to Database Access.

Reply:
Yeah, I can definitely see the argument of changing this to `matchNamespace && (matchLabels ||matchUser)` for the sake of consistency if nothing else.

I'll likely go ahead and change this, but I'm really curious if this was a deliberate design decision originally. The reason I'm a bit perplexed by this is that deny rules are basically calculated differently than allow rules and I couldn't find any reference in the docs/code (apart from the actual code 😄) that mention this behavior is intentional. @klizhentas Do you remember by chance?

Reply:
@russjones I've updated it now so deny rules also match greedily and updated the tests accordingly. I also refactored these `CheckAccessToDatabase` functions a bit and instead of having 3 separate ones for different scenarios, was able to reduce to just a single method that accepts a list of Matchers: https://github.com/gravitational/teleport/pull/5453/commits/cd482301c3a4334df205a29a35c80d49cb8dec2f.

Reply:
I agree that `deny` rules should be greedy with `||` both for backwards compatibility and greedy blocking purposes (better to block more, not less in case of deny rules)